games-roguelike/FTL-gog: new package

kitsunenokenja opened the pull request:
With this PR I'm looking to have a myriad of questions answered. Assuming it's acceptable to add ebuilds for GOG-based applications, here are details on assumptions made in the ebuild.

`LICENSE`: there is apparently no explicit licence provided aside from those that cover the code that GOG uses for its own installer, independent from the game it's installing. The `all-rights-reserved` licence in the repo looks like a good candidate to cover this as a proprietary non-distributable application.

`src_unpack` cleanly prepares the payload by stripping the shell script portion prefixed to the binary payload which is just a ZIP archive. Although this is preparing the inherit call to unzip cleanly, it effectively duplicates the distfile and for some GOG items, this can consume quite a bit of filesystem space. It seems that `unzip` will tolerate the useless bytes in front of the actual archive and still decompress anyway, with a warning echoed. However, since we need to call unpack in an ebuild it may not identify the file as a ZIP archive alone.

`src_prepare` discards some useless stuff as indicated by the comment. This also raises the question for this ebuild specifically: because this game ships with both 32-bit and 64-bit binaries, should we make use of the `abi_x86_32` USE flag to identify which the user wants to keep and let the ebuild discard the other? The scripts packaged with FTL will choose on their own based on the CPU arch it's running on if we don't customise this further.

If this is improved and accepted into the repo, I will use the details gathered here to prepare other GOG-based ebuild contributions using the `-gog` suffix in the package names, and always installing to `/opt/gog` which can serve as a namespace to avoid conflicts with other ebuilds. (We already have `games-roguelike/FTL` but it is not compatible at all if the source payload is from GOG.)

diogocp commented:
> Assuming it's acceptable to add ebuilds for GOG-based applications

I submitted one myself and it was accepted ([games-rpg/baldurs-gate-ee](https://github.com/gentoo/gentoo/tree/master/games-rpg/baldurs-gate-ee)). Take a look at the ebuild, it answers some of your questions. I've also prepared an [ebuild for Torchlight II](https://github.com/diogocp/gentoo/tree/torchlight2/games-rpg/torchlight2) which deals with the 32-/64-bit binaries question, but I haven't submitted it yet.

diogocp commented on games-roguelike/FTL-gog/FTL-gog-1.6.3.ebuild:
```diff
@@ -0,0 +1,57 @@
+	sys-libs/zlib
+	virtual/opengl"
+
+S="${WORKDIR}"
```
Use `S="${WORKDIR}/data/noarch"` instead, to simplify the paths in `src_install`.

diogocp commented on games-roguelike/FTL-gog/FTL-gog-1.6.3.ebuild:
```diff
@@ -0,0 +1,57 @@
+	exeinto /opt/gog/FTL/game/data
+	doexe data/noarch/game/data/FTL*
+
+	make_desktop_entry /opt/gog/FTL/start.sh "FTL: Advanced Edition" /usr/share/pixmaps/FTL.png
```
Add at least the "Game" category. See the [main categories](https://standards.freedesktop.org/menu-spec/latest/apa.html#main-category-registry) and [additional categories](https://standards.freedesktop.org/menu-spec/latest/apas02.html).

diogocp commented on games-roguelike/FTL-gog/FTL-gog-1.6.3.ebuild:
```diff
@@ -0,0 +1,57 @@
+
+src_install() {
+	insinto /opt/gog/FTL
+	doins -r data/noarch/*
```
I don't have this installer to check, but at least in Baldur's Gate, I just installed `data/noarch/game` into `/opt`. The manuals go into the docdir, and the support stuff was not useful.

If you did that, you could drop your `src_prepare`.

diogocp commented on games-roguelike/FTL-gog/FTL-gog-1.6.3.ebuild:
```diff
@@ -0,0 +1,57 @@
+	exeinto /opt/gog/FTL/game
+	doexe data/noarch/game/FTL
+	exeinto /opt/gog/FTL/game/data
+	doexe data/noarch/game/data/FTL*
```
These 6 lines could be simplified to a single `fperms +x ...`.

kitsunenokenja commented:
@diogocp, thanks for the sample references and suggestions! I was glad to see there is a routine `fperms` to simplify that overly verbose segment I wrote since I did not know about that call beforehand.

For the destination path, if the idea of using `/opt/gog` is deemed unnecessary then it might be wise to put a blocking dependency on the non-GOG package of the same game, and in turn update the ebuild for that version to block this one. As for 32/64 selection, are the use keywords `x86` and `amd64` automatically visible implicitly from the `ABI_X86` env value?

diogocp commented:
> As for 32/64 selection, are the use keywords x86 and amd64 automatically visible implicitly from the ABI_X86 env value?

They are "[ARCH USE Flags](https://devmanual.gentoo.org/general-concepts/use-flags/index.html)", and yes, they are automatically available. I don't they they are related to ABI_X86, which is only used in multilib ebuilds.

diogocp commented on games-roguelike/FTL-gog/FTL-gog-1.6.3.ebuild:
```diff
@@ -0,0 +1,51 @@
+	exeinto /opt/gog/FTL/game
+	doexe game/FTL
+	exeinto /opt/gog/FTL/game/data
+	doexe game/data/FTL*
```
Forgot to delete these lines.

diogocp commented on games-roguelike/FTL-gog/FTL-gog-1.6.3.ebuild:
```diff
@@ -0,0 +1,51 @@
+src_install() {
+	insinto /opt/gog/FTL
+	doins -r *
+	fperms +x start.sh game/FTL game/data/FTL*
```
You need to specify the full path to the files (e.g. `/opt/gog/FTL/start.sh`). Otherwise, this will change the permissions of the files in `${S}`.

diogocp replied:
And I recommend putting the path in a local (e.g. `local dir="/opt/gog/FTL"`), so that you don't repeat yourself and can easily change it in one place.

kitsunenokenja commented:
They're based on ARCH, great, I've incorporated that. Also there is no need to explicitly define the category for the desktop entry creation since it will deduce correctly anyway based on the category of the package atom. Tried using bash nuances for the fperms statement but none of them would work; ended up being a little more verbose as a result. If you know a way to condense it, I'd be glad to revise it again.

Tossed the redundant old lines from the first revision. Thought I had dropped those last time.

kitsunenokenja commented:
Usage of quotes was the issue before. Simplifed and re-emerged, the executable bit applied to all intended targets as expected.

diogocp commented on games-roguelike/FTL-gog/FTL-gog-1.6.3.ebuild:
```diff
@@ -0,0 +1,56 @@
+	fi
+	if ! use amd64; then
+		rm game/data/FTL.amd64 || die
+	fi
```
I think this can be simplified to:
```bash
use x86 || rm game/data/FTL.x86 || die
use amd64 || rm game/data/FTL.amd64 || die
```

kitsunenokenja replied:
Had issues with that syntax which I attempted first. Going to try it again to confirm. Even tried using `[ expr ]` syntax and had problems.

mgorny replied:
@diogocp, we don't do double `||`, it's unreadable and unpredictable.

kitsunenokenja commented:
Condensed conditionals. Not sure what was the issue earlier but it's sane now.

diogocp commented:
One last thing: `games-roguelike/FTL` has some bundled libraries, so this package probably has them too. You should check if they can be replaced by the system versions (sometimes they depend on obsolete or modified versions). Check the other FTL ebuild, you can probably copy most of that part with minimal changes.

kitsunenokenja commented:
There are no libraries in this package. Unless I'm mistaken this suggests dynamic linking to dependencies:

```
$ ldd /opt/gog/FTL/game/data/FTL.amd64 
	linux-vdso.so.1 (0x00007fffdf376000)
	libGL.so.1 => /usr/lib64/libGL.so.1 (0x00007f60d2850000)
	libX11.so.6 => /usr/lib64/libX11.so.6 (0x00007f60d252f000)
	libasound.so.2 => /usr/lib64/libasound.so.2 (0x00007f60d226e000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f60d206a000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f60d1e4a000)
	librt.so.1 => /lib64/librt.so.1 (0x00007f60d1c42000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f60d1934000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f60d1585000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f60d2ab9000)
	libexpat.so.1 => /usr/lib64/libexpat.so.1 (0x00007f60d1364000)
	libxcb-dri3.so.0 => /usr/lib64/libxcb-dri3.so.0 (0x00007f60d1161000)
	libxcb-xfixes.so.0 => /usr/lib64/libxcb-xfixes.so.0 (0x00007f60d0f59000)
	libxcb-present.so.0 => /usr/lib64/libxcb-present.so.0 (0x00007f60d0d56000)
	libxcb-sync.so.1 => /usr/lib64/libxcb-sync.so.1 (0x00007f60d0b4f000)
	libxshmfence.so.1 => /usr/lib64/libxshmfence.so.1 (0x00007f60d094d000)
	libglapi.so.0 => /usr/lib64/libglapi.so.0 (0x00007f60d071d000)
	libXext.so.6 => /usr/lib64/libXext.so.6 (0x00007f60d050c000)
	libXdamage.so.1 => /usr/lib64/libXdamage.so.1 (0x00007f60d0309000)
	libXfixes.so.3 => /usr/lib64/libXfixes.so.3 (0x00007f60d0103000)
	libX11-xcb.so.1 => /usr/lib64/libX11-xcb.so.1 (0x00007f60cff01000)
	libxcb.so.1 => /usr/lib64/libxcb.so.1 (0x00007f60cfcda000)
	libxcb-glx.so.0 => /usr/lib64/libxcb-glx.so.0 (0x00007f60cfabf000)
	libxcb-dri2.so.0 => /usr/lib64/libxcb-dri2.so.0 (0x00007f60cf8ba000)
	libXxf86vm.so.1 => /usr/lib64/libXxf86vm.so.1 (0x00007f60cf6b5000)
	libdrm.so.2 => /usr/lib64/libdrm.so.2 (0x00007f60cf4a5000)
	libXau.so.6 => /usr/lib64/libXau.so.6 (0x00007f60cf2a2000)
	libXdmcp.so.6 => /usr/lib64/libXdmcp.so.6 (0x00007f60cf09c000)
	libbsd.so.0 => /usr/lib64/libbsd.so.0 (0x00007f60cee8a000)
```

diogocp commented:
> Unless I'm mistaken this suggests dynamic linking to dependencies

Yes, but `start.sh` may set LD_LIBRARY_PATH so that your system libraries are not used.

It's weird that it doesn't bundle any libraries, but heh, I'm not complaining.

mgorny commented on games-roguelike/FTL-gog/FTL-gog-1.6.3.ebuild:
```diff
@@ -0,0 +1,52 @@
+pkg_nofetch() {
+	elog "Please buy and download ${SRC_URI} from"
+	elog "https://www.gog.com/game/faster_than_light"
+	elog "and copy it to ${DISTDIR}"
```
DISTDIR is forbidden in `pkg_nofetch`.

kitsunenokenja replied:
What is the preferred practice in this situation then? Should the same suggestion remain but simply without a reference to what is normally `/usr/portage/distfiles` ? This elog section was based on a set precedent in another ebuild.

mgorny replied:
Just say 'your distfiles directory' or something like that. And yes, ebuilds do that a lot and modern Portage versions output completely useless directories which is not helpful.

kitsunenokenja replied:
I had simply dropped the third `elog` line from the ebuild file. I can put the substitute text back in though. If we need another revision I'll put this in with it.

mgorny commented on games-roguelike/FTL-gog/FTL-gog-1.6.3.ebuild:
```diff
@@ -0,0 +1,52 @@
+}
+
+src_unpack() {
+	unpack_zip "${DISTDIR}/${A}"
```
Any reason you're using this special function instead of the default `unpack` function that doesn't require any eclasses?

kitsunenokenja replied:
Yes, there is a reason. The payload is a shell script that is prefixed to a binary payload, which is really just a ZIP archive. If you notice some comments above from an earlier discussion, I had originally written a call to dd to separate the shell script portion from the ZIP archive, since `unzip` will throw a warning about the unnecessary bytes leading up to the actual binary portion, and it was suggested to simply use this unpack call which doesn't seem to concern itself with the small aforementioned infringement.

mgorny replied:
Apparently the only difference is that it passes `-q` to silence the output. Looks like a really big cannon to silence a harmless informational output (and at the same time lose any valuable output there could be).

mgorny commented on games-roguelike/FTL-gog/FTL-gog-1.6.3.ebuild:
```diff
@@ -0,0 +1,52 @@
+
+src_prepare() {
+	default
+	rm -r support/{xdg*,*.{sh,txt}}
```
`|| die`

mgorny commented on games-roguelike/FTL-gog/FTL-gog-1.6.3.ebuild:
```diff
@@ -0,0 +1,52 @@
+src_prepare() {
+	default
+	rm -r support/{xdg*,*.{sh,txt}}
+	use x86 || rm game/data/FTL.x86 || die
```
Please don't do multiple `||` chains, they're unpredictable. Use `if ...` instead.

mgorny commented on games-roguelike/FTL-gog/FTL-gog-1.6.3.ebuild:
```diff
@@ -0,0 +1,52 @@
+}
+
+src_install() {
+	local INSTALLDIR=/opt/gog/FTL
```
Please don't create local variables like this. Or if you really must, local variables should be lowercase per common convention.

kitsunenokenja replied:
Reading other ebuilds is teaching bad habits it seems...

mgorny commented on games-roguelike/FTL-gog/FTL-gog-1.6.3.ebuild:
```diff
@@ -0,0 +1,52 @@
+src_install() {
+	local INSTALLDIR=/opt/gog/FTL
+	insinto "${INSTALLDIR}"
+	doins -r *
```
`doins -r .` to avoid unnecessary glob.

kitsunenokenja replied:
Subtle, but an awesome suggestion. A shame that the `doins -r *` pattern seems to be more common among ebuilds.

mgorny commented on games-roguelike/FTL-gog/FTL-gog-1.6.3.ebuild:
```diff
@@ -0,0 +1,52 @@
+	doins -r *
+	fperms +x "${INSTALLDIR}"/{start.sh,game/{FTL,data/FTL*}}
+
+	make_desktop_entry "${INSTALLDIR}/start.sh" "FTL: Advanced Edition" /usr/share/pixmaps/FTL.png
```
I'm pretty sure you're not supposed to use absolute paths in desktop files.

mgorny commented on games-roguelike/FTL-gog/metadata.xml:
```diff
@@ -0,0 +1,16 @@
+		<name>Rob Levitsky</name>
+	</maintainer>
+	<maintainer type="project">
+		<email>games@gentoo.org</email>
```
@gentoo/games, do you approve?

chewi commented:
I haven't looked closely at this yet but I was wondering about the choice of making this a separate package. I would normally prefer adding this to the existing package, which you can do with USE flags, as I imagine the GOG version isn't really any different apart from how it's distributed? This is what we have traditionally done with games available from different sources, for example games-fps/duke3d-data has `cdinstall` and `gog` USE flags. There is a counter argument though, which is that if different sources don't all carry the latest version then this may confuse users as Portage will try to upgrade to a version they cannot fetch.

I'm also considering having Portage fetch GOG games using lgogdownloader, possibly via `src_unpack` in an eclass to start with, and then via `src_fetch` if I can get it into PMS and Portage. I don't yet know whether this would mean that GOG versions would always need to be "live" (i.e. 9999). No ETA on any of this.

kitsunenokenja commented:
If you compare this ebuild to the original that is not GOG, there are so many differences that it would be a preposterous resemblence to read the ebuild you'd have as a result trying to fuse them together. Despite being the same software, the ebuilds have little to do with one another. The nature of the changes that would have to be made would be analogous to those that would be done if one were to fuse two entirely unrelated ebuilds together into one.

I definitely agree on having an eclass to address all the common logic that any GOG-derived installer source would need. There are at least a dozen ebuilds I could prepare after this, and an eclass would be valuable to have.

diogocp commented on games-roguelike/FTL-gog/FTL-gog-1.6.3.ebuild:
```diff
@@ -0,0 +1,54 @@
+src_prepare() {
+	default
+	rm -r support/{xdg*,*.{sh,txt}} || die
+	if [ use x86 ]; then
```
Should be the other way around: `if ! use x86` ;

kitsunenokenja replied:
Thanks, got it. For some reason this led to `fperms` failing, so I replaced +x with 755, and that works.

diogocp commented:
@kitsunenokenja
> If you compare this ebuild to the original that is not GOG, there are so many differences

Actually they are almost exactly the same. It's just copying the files to /opt, creating a wrapper and a .desktop file. The only difference is that the other package has some bundled libs.

It may be hard to create a unified ebuild unless you own both versions, though.

kitsunenokenja commented:
Exactly, you need to have both to be sure, and I suspect the files are laid out quite differently to the extent that you'd have nothing but if-else sequences making the fusion of the two counter-intuitive. There's also another aspect to consider: by trying to modify that version instead, it would be an inherited change to the existing maintainer who probably can't act upon those changes due to not owning both versions.

mgorny commented on games-roguelike/FTL-gog/FTL-gog-1.6.3.ebuild:
```diff
@@ -0,0 +1,54 @@
+src_prepare() {
+	default
+	rm -r support/{xdg*,*.{sh,txt}} || die
+	if [ ! use x86 ]; then
```
This doesn't work the way you expect it to. Remove the brackets.

kitsunenokenja replied:
Must be a habit of all the if's I've written in bash lately. Brackets stripped.
